src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c

```diff
@@ -1137,33 +1180,118 @@ int ADUC_Communication_Channel_MQTT_Publish(
- * @brief
-*/
+ * @brief Subscribes to a topic using version 5 of the MQTT protocol.
+ * @param[in] commHandle The handle to the communication channel.
+ * @param[in] topic The topic to subscribe to.
+ * @param[out] mid Pointer to an integer where the message ID will be stored. Can be NULL.
+ * @param[in] qos Quality of Service level for the message. Valid values are 0, 1, or 2.
+ * @param[in] options Subscription options. Valid values are 0 or 1.
+ * @param[in] props MQTT v5 properties to be included in the message. Can be NULL if no properties are to be sent.
+ * @param[in] userData Pointer to user-specific data.
+ * @param[in] callback Pointer to the callback function to be invoked when the subscription is complete.
+ * @return MOSQ_ERR_SUCCESS upon successful completion.
+ */
 int ADUC_Communication_Channel_MQTT_Subscribe(
     ADUC_AGENT_MODULE_HANDLE commHandle,
     const char* topic,
-
     int* mid,
     int qos,
     int options,
-    const mosquitto_property* props)
+    const mosquitto_property* props,
+    void* userData,
+    MQTT_ON_SUBSCRIBE_V5_CALLBACK callback)
 {
-    ADUC_AGENT_MODULE_INTERFACE* interface = (ADUC_AGENT_MODULE_INTERFACE*)commHandle;
+    if (commHandle == NULL || topic == NULL)
+    {
+        Log_Error("Null parameter (commHandle=%p, topic=%p)", commHandle, topic);
+        return MOSQ_ERR_INVAL;
+    }
 
-    if (interface == NULL)
+    int mosqResult = MOSQ_ERR_UNKNOWN;
+    ADU_MQTT_COMMUNICATION_MGR_STATE* commMgrState = CommunicationManagerStateFromModuleHandle(commHandle);
+    ADUC_MQTT_SUBSCRIBE_CALLBACK_INFO* callbackInfo = NULL;
+
+    do
     {
-        Log_Error("Null parameter (interface=%p)", interface);
-        return -1;
+        // Create a tracking data.
+        callbackInfo = (ADUC_MQTT_SUBSCRIBE_CALLBACK_INFO*)malloc(sizeof(ADUC_MQTT_SUBSCRIBE_CALLBACK_INFO));
+        if (callbackInfo == NULL)
+        {
+            Log_Error("No mem. can't verify the subscription. (topic%s)", topic);
+            mosqResult = MOSQ_ERR_NOMEM;
+            break;
+        }
+
+        callbackInfo->messageId = -1;
+        if (mallocAndStrcpy_s(&callbackInfo->topic, topic) != 0)
+        {
+            Log_Error("No mem. can't copy the topic (topic%s)", topic);
+            mosqResult = MOSQ_ERR_NOMEM;
+            break;
+        }
+
+        callbackInfo->callback = callback;
+        callbackInfo->userData = userData;
+
+        // Add the tracking data to the list.
+        VECTOR_push_back(commMgrState->pendingSubscriptions, callbackInfo, 1);
+
+        mosqResult =
+            mosquitto_subscribe_v5(commMgrState->mqttClient, &callbackInfo->messageId, topic, qos, options, props);
+        if (mosqResult != MOSQ_ERR_SUCCESS)
+        {
+            Log_Error("mosquitto_subscribe_v5 failed (mosqRes:%d)", mosqResult);
+            break;
+        }
+
+        Log_Info("Subscribing to topic: %s, mid: %d", topic, callbackInfo->messageId);
+        mosqResult = MOSQ_ERR_SUCCESS;
+    } while (false);
```

Comment:
do { .. } while (false); screams it should be in a separate function, so let's not do this.  You can also just create a normal scope with { ... }, but a named function would be better.